pkg/controller/syncsetinstance/syncsetinstance_controller.go

```diff
@@ -0,0 +1,714 @@
+			return reconcile.Result{}, nil
+		}
+		return r.syncDeletedSyncSetInstance(ssi, ssiLog)
+	}
+
+	if !controllerutils.HasFinalizer(ssi, hivev1.FinalizerSyncSetInstance) {
+		ssiLog.Debugf("adding  finalizer")
+		if err := r.addSyncSetInstanceFinalizer(ssi); err != nil {
+			ssiLog.WithError(err).Error("error adding finalizer")
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	cd := &hivev1.ClusterDeployment{}
+	cdName := types.NamespacedName{Namespace: ssi.Namespace, Name: ssi.Spec.ClusterDeployment.Name}
+	ssiLog = ssiLog.WithField("clusterdeployment", cdName)
+	err = r.Get(context.TODO(), cdName, cd)
+	if err != nil {
+		ssiLog.WithError(err).Error("error looking up clusterdeployment")
+		return reconcile.Result{}, nil
+	}
+	if !cd.DeletionTimestamp.IsZero() {
+		ssiLog.Debug("clusterdeployment is being deleted")
+		return reconcile.Result{}, nil
+	}
+
+	if !cd.Status.Installed {
+		ssiLog.Debug("cluster installation is not complete")
+		return reconcile.Result{}, nil
+	}
+
+	if len(cd.Status.AdminKubeconfigSecret.Name) == 0 {
+		ssiLog.Debug("admin kubeconfig secret name is not set on clusterdeployment")
+		return reconcile.Result{}, nil
+	}
+
+	ssiLog.Info("reconciling syncsetinstance")
+	spec, err := r.getSyncSetCommonSpec(ssi, ssiLog)
+	if spec == nil || err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// get kubeconfig for the cluster
+	adminKubeconfigSecret := &corev1.Secret{}
+	err = r.Get(context.TODO(), types.NamespacedName{Name: cd.Status.AdminKubeconfigSecret.Name, Namespace: cd.Namespace}, adminKubeconfigSecret)
+	if err != nil {
+		ssiLog.WithError(err).Error("unable to load admin kubeconfig secret")
+		return reconcile.Result{}, err
+	}
+	kubeConfig, err := controllerutils.FixupKubeconfigSecretData(adminKubeconfigSecret.Data)
+	if err != nil {
+		ssiLog.WithError(err).Error("unable to fixup cluster client")
+		return reconcile.Result{}, err
+	}
+	dynamicClient, err := r.dynamicClientBuilder(string(kubeConfig))
+	if err != nil {
+		ssiLog.WithError(err).Error("unable to build dynamic client")
+		return reconcile.Result{}, err
+	}
+
+	ssiLog.Debug("applying sync set")
+	original := ssi.DeepCopy()
+	applier := r.applierBuilder(kubeConfig, ssiLog)
+	applyErr := r.applySyncSet(ssi, spec, dynamicClient, applier, kubeConfig, ssiLog)
+	err = r.updateSyncSetInstanceStatus(ssi, original, ssiLog)
+	if err != nil {
+		ssiLog.WithError(err).Errorf("error updating syncsetinstance status")
+		return reconcile.Result{}, err
+	}
+
+	ssiLog.Info("done reconciling syncsetinstance")
+	return reconcile.Result{}, applyErr
+}
+
+func (r *ReconcileSyncSetInstance) addSyncSetInstanceFinalizer(ssi *hivev1.SyncSetInstance) error {
+	controllerutils.AddFinalizer(ssi, hivev1.FinalizerSyncSetInstance)
+	return r.Update(context.TODO(), ssi)
+}
+
+func (r *ReconcileSyncSetInstance) syncDeletedSyncSetInstance(ssi *hivev1.SyncSetInstance, ssiLog log.FieldLogger) (reconcile.Result, error) {
```

Comment:
Is the finalizer needed at all?

Reply:
hmm... so I could rework the way I handle not finding the corresponding selectorsyncset/syncset (right now I do a delete, then handle the deleted syncsetinstance). I think it could work. Will update.

Reply:
I assume finalizer proved useful in the end?